src/Disks/ObjectStorages/ObjectStorageVFSGCThread.cpp

```diff
@@ -0,0 +1,190 @@
+    VFSSnapshotWithObsoleteObjects out;
+
+    Coordination::Requests requests(log_batch_length);
+    for (size_t i = 0; i < log_batch_length; ++i)
+        requests[i] = zkutil::makeGetRequest(getNode(start_logpointer + i));
+
+    std::vector<VFSTransactionLogItem> log_batch(log_batch_length);
+    VFSTransactionLogItem previous_snapshot_log_item;
+
+    /// Precondition: when we write a snapshot on a previous replica, the snapshot writing operation is
+    /// put in log, so when we process next batch, we can get the snapshot remote path from log. Then we
+    /// construct a StoredObject and read directly from it.
+    const Coordination::Responses responses = storage.zookeeper->multi(requests);
+
+    for (size_t i = 0; i < log_batch_length; ++i)
+    {
+        const String & log_item_str = dynamic_cast<const Coordination::GetResponse &>(*responses[i]).data;
+        auto log_item = VFSTransactionLogItem::deserialize(log_item_str);
+
+        if (log_item.type == CreateInode && log_item.local_path.starts_with(VFS_SNAPSHOT_PREFIX))
+        {
+            if (!previous_snapshot_log_item.remote_path.empty())
+                throw Exception(
+                    ErrorCodes::LOGICAL_ERROR,
+                    "More than one snapshot entry ({}, {}) found for log batch [{};{}]",
+                    previous_snapshot_log_item,
+                    log_item,
+                    start_logpointer,
+                    end_logpointer);
+
+            previous_snapshot_log_item = log_item;
+        }
+
+        log_batch[i] = std::move(log_item);
+    }
+
+    if (has_previous_snapshot)
+    {
+        // TODO myrrc what if replica capturing log entry doesn't get snapshot due to stale read?
+        // Consider using zk->sync(VFS_LOG_BASE_NODE)
+        if (previous_snapshot_log_item.remote_path.empty())
+            throw Exception(
+                ErrorCodes::LOGICAL_ERROR,
+                "No snapshot for {} found in log entries [{};{}]",
+                start_logpointer - 1,
+                start_logpointer,
+                end_logpointer);
+        const size_t snapshot_logpointer
+            = parseFromString<size_t>(previous_snapshot_log_item.local_path.substr(VFS_SNAPSHOT_PREFIX.size()));
+        if (snapshot_logpointer + 1 != start_logpointer)
+            throw Exception(
+                ErrorCodes::LOGICAL_ERROR, "Snapshot logpointer {} != start logpointer {}", snapshot_logpointer, start_logpointer);
+
+        previous_snapshot_log_item.type = Unlink;
+        // Issue previous snapshot remote file for removal (no local metadata file as we use direct readObject)
+        log_batch.emplace_back(previous_snapshot_log_item);
+
+        auto snapshot_buf = storage.object_storage->readObject(previous_snapshot_log_item);
+        auto snapshot_compressed_buf = CompressedReadBuffer{*snapshot_buf};
+        String snapshot_str;
+        readStringUntilEOF(snapshot_str, snapshot_compressed_buf);
```

Comment:
snapshot in RAM... well it should not be too big normally (but hard to be sure)

Possible alternative - have data in snapshot sorted (by paths), sort also the log by path and time, then they can be merged in a streaming manner with a tiny memory footprint.